flang/docs/OpenMP-descriptor-management.md

```diff
@@ -0,0 +1,125 @@
+the container. As an example, a small C++ program that is equivalent to the concept described, with the 
+`mock_descriptor` class being representative of the class utilised for descriptors in Clang:
+
+```C++
+struct mock_descriptor {
+  long int x;
+  std::byte x1, x2, x3, x4;
+  void *pointer;
+  long int lx[1][3];
+};
+
+int main() {
+mock_descriptor data;
+#pragma omp target map(tofrom: data, data.pointer[:upper_bound])
+{
+    do something... 
+}
+
+ return 0;
+}
+```
+
+In the above, we have to map both the containing structure, with its non-pointer members and the
+data pointed to by the pointer contained within the structure to appropriately access the data. This 
+is effectively what is done with descriptor types for the time being. Other pointers that are part 
+of the descriptor container such as the addendum should also be treated as the data pointer is 
+treated.
+
+Currently, Flang will lower these descriptor types in the OpenMP lowering (lower/OpenMP.cpp) similarly
+to all other map types, generating an omp.MapInfoOp containing relevant information required for lowering
+the OpenMP dialect to LLVM-IR during the final stages of the MLIR lowering. However, after 
+the lowering to FIR/HLFIR has been performed an OpenMP dialect specific pass for Fortran, 
+`OMPDescriptorMapInfoGenPass` (Optimizer/OMPDescriptorMapInfoGen.cpp) will expand the 
+`omp.MapInfoOp`'s containing descriptors (which currently will be a `BoxType` or `BoxAddrOp`) into multiple 
+mappings, with one extra per pointer member in the descriptor that is supported on top of the original
+descriptor map operation. These pointers members are linked to the parent descriptor by adding them to 
+the member field of the original descriptor map operation, they are then inserted into the relevant map
+owning operation's (`omp.TargetOp`, `omp.DataOp` etc.) map operand list and in cases where the owning operation
+is `IsolatedFromAbove`, it also inserts them as `BlockArgs` to canonicalize the mappings and simplify lowering.
+
+An example transformation by the `OMPDescriptorMapInfoGenPass`:
+
+```
+
+...
+%12 = omp.map_info var_ptr(%1#1 : !fir.ref<!fir.box<!fir.ptr<!fir.array<?xi32>>>>, !fir.box<!fir.ptr<!fir.array<?xi32>>>) map_clauses(tofrom) capture(ByRef) bounds(%11) -> !fir.ref<!fir.box<!fir.ptr<!fir.array<?xi32>>>> {name = "arg_alloc"}
+...
+omp.target map_entries(%12 -> %arg1, %13 -> %arg2 : !fir.ref<!fir.box<!fir.ptr<!fir.array<?xi32>>>>, !fir.ref<i32>) {
+    ^bb0(%arg1: !fir.ref<!fir.box<!fir.ptr<!fir.array<?xi32>>>>, %arg2: !fir.ref<i32>):
+...
+
+====>
+
+...
+%12 = fir.box_offset %1#1 base_addr : (!fir.ref<!fir.box<!fir.ptr<!fir.array<?xi32>>>>) -> !fir.llvm_ptr<!fir.ref<!fir.array<?xi32>>>
+%13 = omp.map_info var_ptr(%12 : !fir.llvm_ptr<!fir.ref<!fir.array<?xi32>>>, !fir.array<?xi32>) map_clauses(tofrom) capture(ByRef) bounds(%11) -> !fir.llvm_ptr<!fir.ref<!fir.array<?xi32>>> {name = ""}
```

Comment:
Why is the var_ptr a pointer to a pointer? This inconsistency feels unintuitive to me. A mapping operation is intended to map the data pointed to by a pointer recorded in `var_ptr`. In this case, it feels it is mapping a pointer instead of the data itself.

Reply:
If I am understanding correctly (which I might not be, I'm sorry if that's the case), I believe that's a side affect of how the box_offset works currently (https://github.com/llvm/llvm-project/blob/main/flang/test/Fir/box-offset.fir#L16) which gets the address of the pointer fields in the descriptor, which at least currently is required for us to access the relevant descriptor fields on lowering without needing specialised descriptor related information (resolves to a GEP of the field in the descriptor structure if I recall correctly) stored in the map_info operation. I believe in this particular example (which in hindsight is perhaps not the clearest example for the document) we are also taking the address of an intermediate allocation for an assumed shape or sized argument, which is necessary for the box_addr to be utilised (requires an in memory box), which may also cloud things a little further.

Reply:
> If I am understanding correctly (which I might not be, I'm sorry if that's the case), I believe that's a side affect of how the box_offset works currently (https://github.com/llvm/llvm-project/blob/main/flang/test/Fir/box-offset.fir#L16) which gets the address of the pointer fields in the descriptor, which at least currently is required for us to access the relevant descriptor fields on lowering without needing specialised descriptor related information (resolves to a GEP of the field in the descriptor structure if I recall correctly) stored in the map_info operation. I believe in this particular example (which in hindsight is perhaps not the clearest example for the document) we are also taking the address of an intermediate allocation for an assumed shape or sized argument, which is necessary for the box_addr to be utilised (requires an in memory box), which may also cloud things a little further.

Everything you said sounds right to me. And the fact that you need to create storage in some cases to be able to take address of box also sounds right to me. It's not just fir.box_offset that requires "pointer" type, but the data operations in both acc and omp dialects.

That said, the main concern I have is that you are storing a pointer to a pointer in a spot where it is not expected (semantics of var_ptr says `The address of variable to copy` and not `the address of the address`). I guess what I am saying is, should the result of fir.box_offset be stored in var_ptr_ptr instead? That is where I intended it to be used in the acc dialect.

Reply:
That's very true and I agree that it likely is more suitable to the var_ptr_ptr field, but the lowering isn't really setup very well at the moment to distinguish between the both just yet unfortunately, it always utilises the var_ptr currently (although, it could be modified to utilise var_ptr_ptr when it's provided in place of var_ptr, but that seems like a band-aid). I am hoping to change that soon in a future PR, but I am also considering changing the name to base pointer/offload pointer (or perhaps base pointer would be better as a new field) as that's currently what these are used as/most analogous to in the OpenMP lowering (initial seeds for filling in the kernel arg structs, baseptr and offload_ptr arguments).

Reply:
I missed this in the review. Thanks @razvanlupusoru for bringing this up.

Besides the address of address issue, the documentation for the field also says that `var_ptr_ptr` is used when the variable mapped is a member of a structure. Since the descriptor mapping is being set up analogous to structure members it probably makes sense to use it to avoid confusion.
https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td#L1243

```
    - `var_ptr_ptr`: Used when the variable copied is a member of a class, structure
      or derived type and refers to the originating struct.
```

I am OK if this will be changed in follow up patches but please wait for @razvanlupusoru

Reply:
No worries, either fixing it in this PR or delaying it works for me, but as I said. I am considering changing these fields to base pointer and offload pointer for OpenMP in the near future.

Reply:
> (although, it could be modified to utilise var_ptr_ptr when it's provided in place of var_ptr, but that seems like a band-aid).
I am not sure I understand - it doesn't seem like a band-aid to me. The current approach seems like an issue to me where mapping semantics are overloaded and the semantics no longer mean map the data pointed to.

> I am hoping to change that soon in a future PR, but I am also considering changing the name to base pointer/offload pointer (or perhaps base pointer would be better as a new field)
I am curious what you have in mind for this and whether it is more intuitive than the current model. And just to be clear, I prefer convincing in the form that the new IR is easier to understand/optimize, not that it matches LLVM's OpenMPIRBuilder. IMO, the omp dialect should not have coupling with the IR builder that would make it harder to optimize and lower through MLIR's GPU pipeline.

> No worries, either fixing it in this PR or delaying it works for me, but as I said.
I am OK with whatever you choose. The initial fix seems straight forward - just use var_ptr_ptr instead of var_ptr.

Reply:
> > (although, it could be modified to utilise var_ptr_ptr when it's provided in place of var_ptr, but that seems like a band-aid).
> 
> I am not sure I understand - it doesn't seem like a band-aid to me. The current approach seems like an issue to me where mapping semantics are overloaded and the semantics no longer mean map the data pointed to.

In the sense that I'd like to change the semantics of var_ptr/var_ptr_ptr -> base pointer and pointer, it means that I'd be modifying it again in the near future. But it's not a problem in any case.

> > I am hoping to change that soon in a future PR, but I am also considering changing the name to base pointer/offload pointer (or perhaps base pointer would be better as a new field)
> 
> I am curious what you have in mind for this and whether it is more intuitive than the current model. And just to be clear, I prefer convincing in the form that the new IR is easier to understand/optimize, not that it matches LLVM's OpenMPIRBuilder. IMO, the omp dialect should not have coupling with the IR builder that would make it harder to optimize and lower through MLIR's GPU pipeline.
> 

It's easier to understand from the perspective of it is what they're being used as currently when lowering to LLVM-IR, alongside it's also the terminology used by the OpenMP specification, Clang and libomptarget. So I would imagine it'd be a lot clearer to readers and implementers alike what the intent is. In this case I do not see how it'd make it any easier or harder to work with MLIR's GPU pipeline, especially as the current suggested combination of the GPU+OMP dialect seems to utilise the OMPIRBuilder. However, if you think it's clearer as is, that's fine, we can see as we progress.

> > No worries, either fixing it in this PR or delaying it works for me, but as I said.
> 
> I am OK with whatever you choose. The initial fix seems straight forward - just use var_ptr_ptr instead of var_ptr.

Hopefully that should be the case.

Reply:
The most recent commit tries to address your concerns, hopefully it's reasonable, but please do point out if it is not :-)

Reply:
Thank you for all your effort in trying to make reviewers happy! Yes - it is better - at least from a type perspective. I am not sure I completely understand why var_ptr had to be made optional (the data action uses var_ptr to know which "host pointer" needs a device pointer - the var_ptr_ptr is for that offset determination in parent). Possibly because of how LLVM decomposes the actions. But alas, I think at this point you probably need to move this forward. So consider an approval from me - I won't formally approve as I feel that requires more due diligence to look into detail in all parts of your change.

Regarding the renaming of the fields, I can understand your explanation from the point-of-view of consistency with OpenMPIRBuilder (and the OpenMP spec of base pointer). However, I don't necessarily think anything needs removed. Namely each existing piece is useful as far as I can see and base pointer maybe could be in addition.

Namely, the existing var_ptr is more precise than base_pointer since it allows partial mapping. Consider the example in the OpenMP standard for data terminology:
COMMENT: For the array section (*p0).x0[k1].p1->p2[k2].x1[k3].x2[4][0:n], where identiﬁers pi have a pointer type declaration and identiﬁers xi have an array type declaration, the base pointer is: (*p0).x0[k1].p1->p2.
var_ptr can capture the address of "(*p0).x0[k1].p1->p2[k2].x1[k3].x2[4]"

Reply:
>  I'll leave it up to @kiranchandramohan to decide what he wishes done in this case.

If splitting the Flang and MLIR dialect portions and the translation and OpenMPIRBuilder portions into separate patches would help get a review and acceptance from @razvanlupusoru then that would be great. Otherwise, you can go ahead and address concerns in future patches.

Reply:
More than happy to do the split if it helps at all, I'm aware the PR was big to begin with and has only grown. They will all have to land at the same time though and I have a feeling it'll break some buildbots during the time they take to land (couple of minutes at worst, so not a big deal). Just let me know if this is something you'd like to do @razvanlupusoru and if it'd help you feel more capable of giving acceptance.

On a side-note I will likely do a revision of the last commit to move the var_ptr field to var_ptr_ptr, to hopefully affect the operation a little less.

Reply:
Took a slightly different approach to the movement of the base_addr to the var_ptr_ptr field that doesn't require as many changes. Still not too sure what the best replacement for the var_ptr field for it is however unfortunately, but I've opted for the descriptor it comes from for the moment, I am open to other suggestions!

Reply:
Small ping for an answer to @kiranchandramohan's question if possible @razvanlupusoru so that I can move forward with either splitting the PR or landing it next week! 

I have a follow up PR I am hoping to have ready next week that builds on this, so it would be nice to close this one out if at all possible, if it's in a reasonable enough state for that of course.

Reply:
I took another high level look at the implementation and the tests. I also took a look at the details of the dialect changes. Overall looks good to me and it feels that @agozillon addressed all of the concerns/suggestions brought up. So please consider an unofficial approval from me :) To give a formal approval, I would need to spend more time thinking about this - and at this point I think if any concerns arise, they can be addressed in follow-up changes.

Reply:
Thank you very much @razvanlupusoru I'm sorry for the bother. But yes, if any concerns arise with the implementation I am more than happy to make the alterations necessary to alleviate them (and I am still very happy to swap to the OpenACC method for descriptor mapping at some point in the future). 

As it stands this PR adds functionality we do not currently have and some initial record type mapping code. Which I hope to build on further in a PR I hope to have up a little after this lands (there will likely be more than one derived type/record related PR, so this would be the first of a few).

Are we happy enough with the current level of acceptance @kiranchandramohan for me to go ahead and try to land this either later today or tomorrow? I am also happy to give the PR a few more days for people to comment on it if they desire if you wish.